src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/PathNormalizer.cs

```diff
@@ -1,12 +1,128 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.Buffers;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Internal.Http
 {
     public static class PathNormalizer
     {
+        // In-place implementation of the algorithm from https://tools.ietf.org/html/rfc3986#section-5.2.4
+        public static unsafe int RemoveDotSegments(Span<byte> span)
+        {
+            var length = span.Length;
+
+            fixed (byte* start = &span.DangerousGetPinnableReference())
+            {
+                var end = start + length;
+                var src = start;
+                var dst = start;
+
+                while (src < end)
+                {
+                    if (end - src >= 3 && *src == '.' && *(src + 1) == '.' && *(src + 2) == '/')
```

Comment:
Also we had chars cached as bytes in Frame would it help here if we compare with const bytes vs chars?